hw3/HaHaFi/multiplication.cpp

```diff
@@ -19,72 +19,59 @@ inline void Check_mutiply_valid(Matrix mat1, Matrix mat2)
     }
 }
 
-
-
 Matrix multiply_naive(Matrix &mat1, Matrix &mat2)
 {
     Check_mutiply_valid(mat1, mat2);
-    Matrix ret_mat(mat1.n_row(), mat2.n_col());
-    for (size_t i = 0; i < mat1.n_row(); i++)
+    const size_t K = mat1.n_col(), M = mat1.n_row(), N = mat2.n_col();
+    Matrix ret_mat(M, N);
+    for (size_t i = 0; i < M; i++)
     {
-        for (size_t j = 0; j < mat2.n_col(); j++)
+        const size_t ret_leader=i * N;
+        for (size_t j = 0; j < N; j++)
         {
             double count = 0;
-            for (size_t k = 0; k < mat1.n_col(); k++)
+            const size_t m1_leader = i * K;
+            for (size_t k = 0; k < K; k++)
             {
-                count += mat1(i, k) * mat2(k, j);
+                count += mat1.m_buffer[m1_leader + k] * mat2.m_buffer[k * N + j];
             }
-            ret_mat(i, j) = count;
+            ret_mat.m_buffer[ret_leader+ j] = count;
         }
```

Comment:
In the navie function, I optimize the index operation, reduce a lot of multiplication.

Reply:
What is the runtime difference between the two implementations?

Reply:
navie function with matrix  (1000 * 1000) @ (1000 * 1000)

Almost three times faster.

Original implementation
```
multiply_naive time below:
min     max     mean    std_dev
3.721   3.823   3.769   0.041
```
Revised implementation
```
multiply_naive time below:
min     max     mean    std_dev
1.175   1.196   1.185   0.010
```